pydra/engine/boutiques.py

```diff
@@ -0,0 +1,221 @@
+            rerun=rerun,
+            **kwargs,
+        )
+        self.strip = strip
+
+    def _download_spec(self, zenodo):
+        """ usind bosh pull to download the zenodo file"""
+        spec_file = (
+            Path(os.environ["HOME"])
+            / ".cache/boutiques/production"
+            / (zenodo.replace(".", "-") + ".json")
+        )
+        for i in range(3):
+            if not spec_file.exists():
+                sp.run(["bosh", "pull", zenodo])
+        if not spec_file.exists():
+            raise Exception(f"can't pull zenodo file {zenodo}")
+        return spec_file
+
+    def _prepare_input_spec(self):
+        """ creating input spec from the zenodo file"""
+        binputs = self.bosh_spec["inputs"]
+        self._input_spec_keys = {}
+        fields = []
+        for input in binputs:
+            name = input["id"]
+            if input["type"] == "File":
+                tp = File
+            elif input["type"] == "String":
+                tp = str
+            elif input["type"] == "Number":
+                tp = float
+            elif input["type"] == "Flag":
+                tp = bool
+            else:
+                tp = None
+            # adding list
+            if tp and "list" in input and input["list"]:
+                tp = ty.List[tp]
+
+            mdata = {
+                "help_string": input.get("description", None) or input["name"],
+                "mandatory": not input["optional"],
+                "argstr": input.get("command-line-flag", None),
+            }
+            fields.append((name, tp, mdata))
+            self._input_spec_keys[input["value-key"]] = "{" + f"{name}" + "}"
+
+        spec = SpecInfo(name="Inputs", fields=fields, bases=(ShellSpec,))
+        return spec
+
+    def _prepare_output_spec(self):
+        """ creating output spec from the zenodo file"""
+        boutputs = self.bosh_spec["output-files"]
+        fields = []
+        for output in boutputs:
+            name = output["id"]
+            path_template = reduce(
+                lambda s, r: s.replace(*r),
+                self._input_spec_keys.items(),
+                output["path-template"],
+            )
+            mdata = {
+                "help_string": output.get("description", None) or output["name"],
+                "mandatory": not output["optional"],
+                "output_file_template": path_template,
+            }
+            fields.append((name, attr.ib(type=File, metadata=mdata)))
+
+        spec = SpecInfo(name="Outputs", fields=fields, bases=(ShellOutSpec,))
+        return spec
+
+    def _command_args_single(self, state_ind, ind=None):
+        """Get command line arguments for a single state"""
+        input_filepath = self._input_file(state_ind=state_ind, ind=ind)
+        cmd_list = (
+            self.inputs.executable + [input_filepath] + self.inputs.args + self.bindings
+        )
+        return cmd_list
+
+    def _input_file(self, state_ind, ind=None):
```

Comment:
this and the previous function appears to interact with state. that seems odd to me. we should not be expecting people to understand state to write a task. thus far all tasks have been somewhat independent of state, and we should keep it that way.

Reply:
don't understand your comment. pydra prepares the files, this is  a private function

Reply:
i did not understand why this function is needed and how to convey to another developer that if they were to write the boutiques task they would need this function. what is the general role of this function? and how would another developer (say a CWL developer) know that this is needed. this seems to depend on `state_ind`.

Reply:
this function creates `invocation` json file with inputs that is needed for `bosh exec launch`. If task has splitter and provides two files for example for `infile` input, I will create two `invocation` files. I'm assuming that user doesn't know that `bosh` requires this file, but just provides `infile`, etc.

Reply:
shouldn't this be the kind of thing that happens inside `run_task`? i.e. the creation of this file is only required when you have the inputs already set. this is not an input to the task itself but to bosh.

Reply:
yes, but this is used by `command_args` property that is used in `run_task`

Reply:
should probably use more specific name

Reply:
i see that `ShellCommandTask` follows a similar pattern. conceptually does `command_args` need to work for tasks with `State`. is it used anywhere in that form?

wouldn't it be easier if these didn't have to think about state for `command_args` and would raise an Exception if called when the Task has state, but would return the appropriate thing when the Task is stateless.

Reply:
can't answer you right now. I didn't see a problem with it, but can rethink this in a different PR

Reply:
and I actually do think that it's nice to have, it helps for debugging